test/unittests/execute_call_depth_test.cpp

```diff
@@ -0,0 +1,327 @@
+    auto instance = instantiate(*module, {{{host_f}, module->typesec[0]}});
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}), Result(1_u32));
+    EXPECT_EQ(recorded_depth, 1);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit - 2), Result(1_u32));
+    EXPECT_EQ(recorded_depth, DepthLimit - 1);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit - 1), Traps());
+    EXPECT_EQ(recorded_depth, -1000);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit), Traps());
+    EXPECT_EQ(recorded_depth, -1000);
+}
+
+TEST(execute_call_depth, call_host_function_calling_wasm_function)
+{
+    // Test for "wasm-host-wasm" sandwich.
+    // The host function is obligated to bump depth and pass it along.
+
+    /* wat2wasm
+    (func $host_f (import "host" "f") (result i32))
+    (func (result i32) (call $host_f))
+    (func $leaf (result i32) (i32.const 1))
+    */
+    const auto wasm = from_hex(
+        "0061736d010000000105016000017f020a0104686f73740166000003030200000a0b02040010000b040041010"
+        "b");
+
+    static int recorded_depth;
+    constexpr auto host_f = [](std::any&, Instance& instance, const Value*, int depth) noexcept {
+        recorded_depth = depth;
+        return fizzy::execute(instance, 2, {}, depth + 1);
+    };
+
+    const auto module = parse(wasm);
+    auto instance = instantiate(*module, {{{host_f}, module->typesec[0]}});
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}), Result(1_u32));
+    EXPECT_EQ(recorded_depth, 1);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit - 3), Result(1_u32));
+    EXPECT_EQ(recorded_depth, DepthLimit - 2);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit - 2), Traps());
+    EXPECT_EQ(recorded_depth, DepthLimit - 1);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit - 1), Traps());
+    EXPECT_EQ(recorded_depth, -1000);
+
+    recorded_depth = -1000;
+    EXPECT_THAT(execute(*instance, 1, {}, DepthLimit), Traps());
+    EXPECT_EQ(recorded_depth, -1000);
+}
+
+
+/// Infinite recursion
+
+TEST(execute_call_depth, execute_internal_function_infinite_recursion)
+{
+    // This execution must always trap.
+    // Number of $f invocations is counted in the imported global $counter.
+
+    /* wat2wasm
+    (global $counter (import "host" "counter") (mut i64))
+    (func $f
+      (global.set $counter (i64.add (global.get $counter) (i64.const 1)))
+      (call $f)
+    )
+    */
+    const auto wasm = from_hex(
+        "0061736d0100000001040160000002110104686f737407636f756e746572037e01030201000a0d010b00230042"
+        "017c240010000b");
+
+    Value counter;
+    auto instance = instantiate(parse(wasm), {}, {}, {}, {{&counter, {ValType::i64, true}}});
+
+    // When starting from depth 0, the $f is expected to be called CallStackLimit times.
```

Comment:
```suggestion
    // When starting from depth 0, the $f is expected to be called DepthLimit times.
```
?

Technically I guess the comment is correct, just confusing to see the expectation refer to a different const.